<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="zh-Hans" xml:lang="zh-Hans"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.550">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>生信初学者教程（癌症转录组学） - 8&nbsp; 数据对象</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./wrangling-batchEffect.html" rel="next">
<link href="./wrangling-processing.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "没有结果",
    "search-matching-documents-text": "匹配的文档",
    "search-copy-link-title": "复制搜索链接",
    "search-hide-matches-text": "隐藏其它匹配结果",
    "search-more-match-text": "更多匹配结果",
    "search-more-matches-text": "更多匹配结果",
    "search-clear-button-title": "清除",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "取消",
    "search-submit-button-title": "提交",
    "search-label": "搜索"
  }
}</script><script>
MathJax ={
  tex: {
    macros: {
      bm: ["{\\boldsymbol #1}",1],
    }
  }
};
</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="展开或折叠侧边栏导航" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./wrangling-collection.html">数据准备</a></li><li class="breadcrumb-item"><a href="./wrangling-objects.html"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">数据对象</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="展开或折叠侧边栏导航" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">生信初学者教程（癌症转录组学）</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="https://github.com/" title="源代码" class="quarto-navigation-tool px-1" aria-label="源代码"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Download" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download"><i class="bi bi-download"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="./_main.pdf">
              <i class="bi bi-bi-file-pdf pe-1"></i>
            Download PDF
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./_main.epub">
              <i class="bi bi-bi-journal pe-1"></i>
            Download ePub
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./_main.docx">
              <i class="bi bi-bi-file-word pe-1"></i>
            Download Docx
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="切换深色模式"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="搜索"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">欢迎</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./preface.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">前言</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">介绍</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">基础知识</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="展开或折叠此栏">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./knowledge-software.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">软件</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./knowledge-R.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">R语言基础</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./knowledge-mathematics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">数学基础</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./knowledge-database.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">数据库</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">数据准备</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="展开或折叠此栏">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./wrangling-collection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">数据收集</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./wrangling-processing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">数据预处理</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./wrangling-objects.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">数据对象</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./wrangling-batchEffect.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">数据校正</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./wrangling-gtsummary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">数据汇总</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">差异分析</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="展开或折叠此栏">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./different-limma.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">差异分析之limma</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./different-volcano.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">差异结果的火山图</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./different-heatmap.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">差异结果的热图</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">功能分析</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="展开或折叠此栏">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./function-GOterm.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">GO富集分析</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./function-KEGGpathway.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">KEGG通路富集分析</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./function-ssgsea.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">ssGSEA富集分析</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">浸润分析</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="展开或折叠此栏">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./infiltration-ImmuCellAI.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">免疫浸润细胞</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./infiltration-Analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">免疫浸润分析</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">标记筛选</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="展开或折叠此栏">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./biomarker-lassoLR.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">LASSO LR</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./biomarker-BorutaRF.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Boruta RF</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./biomarker-RefSVM.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">REF SVM</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./biomarker-intersect.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">交集特征</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./biomarker-validation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">验证特征</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true">
 <span class="menu-text">关联分析</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true" aria-label="展开或折叠此栏">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./association-immunecell.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">关联分析</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" aria-expanded="true">
 <span class="menu-text">单细胞分析</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" aria-expanded="true" aria-label="展开或折叠此栏">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./scRNA-DataProcess.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">单细胞数据处理</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./scRNA-Normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">单细胞数据标准化</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./scRNA-Cluster.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">单细胞聚类分析</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./scRNA-CellType.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">单细胞细胞识别</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./scRNA-CellValidation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">核心特征单细胞表达</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" aria-expanded="true">
 <span class="menu-text">撰写文章</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" aria-expanded="true" aria-label="展开或折叠此栏">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-9" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./paper-structure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">框架</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./paper-results.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">结果</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./paper-methods.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">方法</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./paper-discussion.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">讨论</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./paper-introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">34</span>&nbsp; <span class="chapter-title">介绍</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./paper-others.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">35</span>&nbsp; <span class="chapter-title">结论和摘要</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./paper-manuscript.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">初稿</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-10" aria-expanded="true">
 <span class="menu-text">附录</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-10" aria-expanded="true" aria-label="展开或折叠此栏">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-10" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">补充知识</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">目录</h2>
   
  <ul>
<li>
<a href="#sec-expressionset-LIRIJP" id="toc-sec-expressionset-LIRIJP" class="nav-link active" data-scroll-target="#sec-expressionset-LIRIJP"><span class="header-section-number">8.1</span> LIRI-JP</a>
  <ul class="collapse">
<li><a href="#sec-expressionset-LIRIJP-package" id="toc-sec-expressionset-LIRIJP-package" class="nav-link" data-scroll-target="#sec-expressionset-LIRIJP-package"><span class="header-section-number">8.1.1</span> 加载R包</a></li>
  <li><a href="#sec-expressionset-LIRIJP-impordata" id="toc-sec-expressionset-LIRIJP-impordata" class="nav-link" data-scroll-target="#sec-expressionset-LIRIJP-impordata"><span class="header-section-number">8.1.2</span> 导入数据</a></li>
  <li><a href="#sec-expressionset-LIRIJP-ExpressionSet" id="toc-sec-expressionset-LIRIJP-ExpressionSet" class="nav-link" data-scroll-target="#sec-expressionset-LIRIJP-ExpressionSet"><span class="header-section-number">8.1.3</span> 生成ExpressionSet</a></li>
  <li><a href="#sec-expressionset-LIRIJP-output" id="toc-sec-expressionset-LIRIJP-output" class="nav-link" data-scroll-target="#sec-expressionset-LIRIJP-output"><span class="header-section-number">8.1.4</span> 输出结果</a></li>
  </ul>
</li>
  <li>
<a href="#sec-expressionset-TCGALIHC" id="toc-sec-expressionset-TCGALIHC" class="nav-link" data-scroll-target="#sec-expressionset-TCGALIHC"><span class="header-section-number">8.2</span> TCGA-LIHC</a>
  <ul class="collapse">
<li><a href="#sec-expressionset-TCGALIHC-package" id="toc-sec-expressionset-TCGALIHC-package" class="nav-link" data-scroll-target="#sec-expressionset-TCGALIHC-package"><span class="header-section-number">8.2.1</span> 加载R包</a></li>
  <li><a href="#sec-expressionset-TCGALIHC-importdata" id="toc-sec-expressionset-TCGALIHC-importdata" class="nav-link" data-scroll-target="#sec-expressionset-TCGALIHC-importdata"><span class="header-section-number">8.2.2</span> 导入数据</a></li>
  <li><a href="#sec-expressionset-TCGALIHC-ExpressionSet" id="toc-sec-expressionset-TCGALIHC-ExpressionSet" class="nav-link" data-scroll-target="#sec-expressionset-TCGALIHC-ExpressionSet"><span class="header-section-number">8.2.3</span> 生成ExpressionSet</a></li>
  <li><a href="#sec-expressionset-TCGALIHC-output" id="toc-sec-expressionset-TCGALIHC-output" class="nav-link" data-scroll-target="#sec-expressionset-TCGALIHC-output"><span class="header-section-number">8.2.4</span> 输出结果</a></li>
  </ul>
</li>
  <li>
<a href="#sec-expressionset-GSE14520" id="toc-sec-expressionset-GSE14520" class="nav-link" data-scroll-target="#sec-expressionset-GSE14520"><span class="header-section-number">8.3</span> GSE14520</a>
  <ul class="collapse">
<li><a href="#sec-expressionset-GSE14520-package" id="toc-sec-expressionset-GSE14520-package" class="nav-link" data-scroll-target="#sec-expressionset-GSE14520-package"><span class="header-section-number">8.3.1</span> 加载R包</a></li>
  <li><a href="#sec-expressionset-GSE14520-importdata" id="toc-sec-expressionset-GSE14520-importdata" class="nav-link" data-scroll-target="#sec-expressionset-GSE14520-importdata"><span class="header-section-number">8.3.2</span> 导入数据</a></li>
  <li><a href="#sec-expressionset-GSE14520-ExpressionSet" id="toc-sec-expressionset-GSE14520-ExpressionSet" class="nav-link" data-scroll-target="#sec-expressionset-GSE14520-ExpressionSet"><span class="header-section-number">8.3.3</span> 生成ExpressionSet</a></li>
  <li><a href="#sec-expressionset-GSE14520-output" id="toc-sec-expressionset-GSE14520-output" class="nav-link" data-scroll-target="#sec-expressionset-GSE14520-output"><span class="header-section-number">8.3.4</span> 输出结果</a></li>
  </ul>
</li>
  <li>
<a href="#sec-expressionset-overlapgene" id="toc-sec-expressionset-overlapgene" class="nav-link" data-scroll-target="#sec-expressionset-overlapgene"><span class="header-section-number">8.4</span> 重叠基因</a>
  <ul class="collapse">
<li><a href="#sec-expressionset-overlapgene-package" id="toc-sec-expressionset-overlapgene-package" class="nav-link" data-scroll-target="#sec-expressionset-overlapgene-package"><span class="header-section-number">8.4.1</span> 加载R包</a></li>
  <li><a href="#sec-expressionset-overlapgene-importdata" id="toc-sec-expressionset-overlapgene-importdata" class="nav-link" data-scroll-target="#sec-expressionset-overlapgene-importdata"><span class="header-section-number">8.4.2</span> 导入数据</a></li>
  <li><a href="#sec-expressionset-overlapgene-common" id="toc-sec-expressionset-overlapgene-common" class="nav-link" data-scroll-target="#sec-expressionset-overlapgene-common"><span class="header-section-number">8.4.3</span> 共有基因</a></li>
  <li><a href="#sec-expressionset-overlapgene-object" id="toc-sec-expressionset-overlapgene-object" class="nav-link" data-scroll-target="#sec-expressionset-overlapgene-object"><span class="header-section-number">8.4.4</span> 生成数据对象</a></li>
  <li><a href="#sec-expressionset-overlapgene-output" id="toc-sec-expressionset-overlapgene-output" class="nav-link" data-scroll-target="#sec-expressionset-overlapgene-output"><span class="header-section-number">8.4.5</span> 输出结果</a></li>
  </ul>
</li>
  <li><a href="#sec-expressionset-summary" id="toc-sec-expressionset-summary" class="nav-link" data-scroll-target="#sec-expressionset-summary"><span class="header-section-number">8.5</span> 总结</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/blob/main/wrangling-objects.qmd" class="toc-action"><i class="bi bi-github"></i>查看代码</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./wrangling-collection.html">数据准备</a></li><li class="breadcrumb-item"><a href="./wrangling-objects.html"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">数据对象</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title"><span id="sec-expressionset" class="quarto-section-identifier"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">数据对象</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><p><em>ExpressionSet</em>是<strong>Biobase</strong><span class="citation" data-cites="huber2015orchestrating">(<a href="references.html#ref-huber2015orchestrating" role="doc-biblioref">Huber 等 2015</a>)</span>包提供的一种数据对象，专为存储生物学实验数据而设计。它允许用户将表型数据、表达谱以及特征数据（如基因或蛋白质标识符）集成在一个统一的数据结构中。在本章节中，将利用这种数据对象来组织并处理获得的三个数据集：<code>LIRI-JP</code>，<code>LIHC-US/TCGA-LIHC</code>和<code>GSE14520</code>。</p>
<section id="sec-expressionset-LIRIJP" class="level2" data-number="8.1"><h2 data-number="8.1" class="anchored" data-anchor-id="sec-expressionset-LIRIJP">
<span class="header-section-number">8.1</span> LIRI-JP</h2>
<p>要生成<code>LIRI-JP</code>的<em>ExpressionSet</em>对象，首先需要加载必要的R包，然后导入数据和元数据，最后将这些数据组合成一个<em>ExpressionSet</em>对象。</p>
<section id="sec-expressionset-LIRIJP-package" class="level3" data-number="8.1.1"><h3 data-number="8.1.1" class="anchored" data-anchor-id="sec-expressionset-LIRIJP-package">
<span class="header-section-number">8.1.1</span> 加载R包</h3>
<p>加载R包：使用<code>rm(list = ls())</code>清空环境内变量</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="fu">library</span>(Biobase)</span>
<span id="cb1-4"><a href="#cb1-4"></a></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="fu">options</span>(<span class="at">stringsAsFactors =</span> F)</span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize =</span> <span class="dv">10000</span> <span class="sc">*</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb1-8"><a href="#cb1-8"></a></span>
<span id="cb1-9"><a href="#cb1-9"></a>grp_names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Early Stage"</span>, <span class="st">"Late Stage"</span>)</span>
<span id="cb1-10"><a href="#cb1-10"></a>grp_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#1F78B4"</span>, <span class="st">"#EE2B2B"</span>)</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="sec-expressionset-LIRIJP-impordata" class="level3" data-number="8.1.2"><h3 data-number="8.1.2" class="anchored" data-anchor-id="sec-expressionset-LIRIJP-impordata">
<span class="header-section-number">8.1.2</span> 导入数据</h3>
<p>输入数据来自于 <a href="wrangling-processing.html" class="quarto-xref"><span>章节 7</span></a></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>LIRIJP_count_data <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"./data/result/profile/LIRI-JP_gene_counts.tsv"</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a>LIRIJP_relative_data <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"./data/result/profile/LIRI-JP_gene_relative.tsv"</span>) </span>
<span id="cb2-3"><a href="#cb2-3"></a>LIRIJP_phenotype <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"./data/result/metadata/LIRI-JP-specimen.csv"</span>)</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="sec-expressionset-LIRIJP-ExpressionSet" class="level3" data-number="8.1.3"><h3 data-number="8.1.3" class="anchored" data-anchor-id="sec-expressionset-LIRIJP-ExpressionSet">
<span class="header-section-number">8.1.3</span> 生成ExpressionSet</h3>
<p>生成<em>ExpressionSet-class</em>数据对象：过滤掉低出现率的基因，默认出现率阈值是<strong>20%</strong></p>
<div class="sourceCode" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>get_ExprSet_LIRIJP <span class="ot">&lt;-</span> <span class="cf">function</span>(</span>
<span id="cb3-2"><a href="#cb3-2"></a>    profile,</span>
<span id="cb3-3"><a href="#cb3-3"></a>    metadata,</span>
<span id="cb3-4"><a href="#cb3-4"></a>    <span class="at">occurrence =</span> <span class="fl">0.2</span>) {</span>
<span id="cb3-5"><a href="#cb3-5"></a>  </span>
<span id="cb3-6"><a href="#cb3-6"></a>  profile_new <span class="ot">&lt;-</span> profile <span class="sc">%&gt;%</span></span>
<span id="cb3-7"><a href="#cb3-7"></a>    tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"V1"</span>)</span>
<span id="cb3-8"><a href="#cb3-8"></a>  metadata_new <span class="ot">&lt;-</span> metadata <span class="sc">%&gt;%</span></span>
<span id="cb3-9"><a href="#cb3-9"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(Tumour_Stage <span class="sc">!=</span> <span class="st">""</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-10"><a href="#cb3-10"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">Group =</span> <span class="fu">case_when</span>(</span>
<span id="cb3-11"><a href="#cb3-11"></a>      Tumour_Stage <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"T1"</span>, <span class="st">"T2"</span>) <span class="sc">~</span> <span class="st">"Early Stage"</span>, </span>
<span id="cb3-12"><a href="#cb3-12"></a>      Tumour_Stage <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"T3"</span>, <span class="st">"T4"</span>) <span class="sc">~</span> <span class="st">"Late Stage"</span>))</span>
<span id="cb3-13"><a href="#cb3-13"></a>  </span>
<span id="cb3-14"><a href="#cb3-14"></a>  <span class="co"># metadata Description</span></span>
<span id="cb3-15"><a href="#cb3-15"></a>  sid <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">intersect</span>(<span class="fu">colnames</span>(profile_new), metadata_new<span class="sc">$</span>SpecimenID)</span>
<span id="cb3-16"><a href="#cb3-16"></a>  phen <span class="ot">&lt;-</span> metadata_new[<span class="fu">pmatch</span>(sid, metadata_new<span class="sc">$</span>SpecimenID), , ]</span>
<span id="cb3-17"><a href="#cb3-17"></a>  phen <span class="ot">&lt;-</span> phen[<span class="fu">pmatch</span>(<span class="fu">unique</span>(phen<span class="sc">$</span>PatientID), phen<span class="sc">$</span>PatientID), , ]</span>
<span id="cb3-18"><a href="#cb3-18"></a>  prof <span class="ot">&lt;-</span> profile_new <span class="sc">%&gt;%</span></span>
<span id="cb3-19"><a href="#cb3-19"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(dplyr<span class="sc">::</span><span class="fu">all_of</span>(phen<span class="sc">$</span>SpecimenID))</span>
<span id="cb3-20"><a href="#cb3-20"></a>  </span>
<span id="cb3-21"><a href="#cb3-21"></a>  <span class="cf">if</span> (<span class="fu">all</span>(<span class="fu">colnames</span>(prof) <span class="sc">==</span> phen<span class="sc">$</span>SpecimenID)) {</span>
<span id="cb3-22"><a href="#cb3-22"></a>    <span class="fu">colnames</span>(prof) <span class="ot">&lt;-</span> phen<span class="sc">$</span>PatientID</span>
<span id="cb3-23"><a href="#cb3-23"></a>  }</span>
<span id="cb3-24"><a href="#cb3-24"></a>  </span>
<span id="cb3-25"><a href="#cb3-25"></a>  <span class="co"># filter features according gene symbol</span></span>
<span id="cb3-26"><a href="#cb3-26"></a>  prof_cln <span class="ot">&lt;-</span> prof <span class="sc">%&gt;%</span></span>
<span id="cb3-27"><a href="#cb3-27"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"Gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-28"><a href="#cb3-28"></a>    <span class="fu">data.frame</span>()</span>
<span id="cb3-29"><a href="#cb3-29"></a>  idx <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"Gene"</span>, <span class="fu">colnames</span>(prof_cln))</span>
<span id="cb3-30"><a href="#cb3-30"></a>  prof_cln<span class="sc">$</span>median <span class="ot">&lt;-</span> <span class="fu">apply</span>(prof_cln[, <span class="sc">-</span>idx], <span class="dv">1</span>, median)</span>
<span id="cb3-31"><a href="#cb3-31"></a>  prof_cln <span class="ot">&lt;-</span> <span class="fu">with</span>(prof_cln, prof_cln[<span class="fu">order</span>(Gene, median, <span class="at">decreasing =</span> T), ])</span>
<span id="cb3-32"><a href="#cb3-32"></a>  prf_deduplicated <span class="ot">&lt;-</span> prof_cln[<span class="sc">!</span><span class="fu">duplicated</span>(prof_cln<span class="sc">$</span>Gene), ] <span class="sc">%&gt;%</span></span>
<span id="cb3-33"><a href="#cb3-33"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>median)</span>
<span id="cb3-34"><a href="#cb3-34"></a>  </span>
<span id="cb3-35"><a href="#cb3-35"></a>  <span class="co"># return result: gene symbol</span></span>
<span id="cb3-36"><a href="#cb3-36"></a>  <span class="fu">rownames</span>(prf_deduplicated) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb3-37"><a href="#cb3-37"></a>  prof_res <span class="ot">&lt;-</span> prf_deduplicated <span class="sc">%&gt;%</span></span>
<span id="cb3-38"><a href="#cb3-38"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(Gene, <span class="fu">everything</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb3-39"><a href="#cb3-39"></a>    tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"Gene"</span>)</span>
<span id="cb3-40"><a href="#cb3-40"></a>  </span>
<span id="cb3-41"><a href="#cb3-41"></a>  <span class="co"># determine the right order between profile and phenotype</span></span>
<span id="cb3-42"><a href="#cb3-42"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(prof_trim)) {</span>
<span id="cb3-43"><a href="#cb3-43"></a>    <span class="cf">if</span> (<span class="sc">!</span>(<span class="fu">colnames</span>(prof_trim)[i] <span class="sc">==</span> <span class="fu">rownames</span>(phen)[i])) {</span>
<span id="cb3-44"><a href="#cb3-44"></a>      <span class="fu">stop</span>(<span class="fu">paste0</span>(i, <span class="st">" Wrong"</span>))</span>
<span id="cb3-45"><a href="#cb3-45"></a>    }</span>
<span id="cb3-46"><a href="#cb3-46"></a>  }</span>
<span id="cb3-47"><a href="#cb3-47"></a>  </span>
<span id="cb3-48"><a href="#cb3-48"></a>  expressionSet <span class="ot">&lt;-</span> <span class="fu">new</span>(<span class="st">"ExpressionSet"</span>, </span>
<span id="cb3-49"><a href="#cb3-49"></a>                       <span class="at">exprs =</span> exprs,</span>
<span id="cb3-50"><a href="#cb3-50"></a>                       <span class="at">phenoData =</span> adf,</span>
<span id="cb3-51"><a href="#cb3-51"></a>                       <span class="at">experimentData =</span> experimentData)</span>
<span id="cb3-52"><a href="#cb3-52"></a>  </span>
<span id="cb3-53"><a href="#cb3-53"></a>  <span class="fu">return</span>(expressionSet)</span>
<span id="cb3-54"><a href="#cb3-54"></a>}</span>
<span id="cb3-55"><a href="#cb3-55"></a></span>
<span id="cb3-56"><a href="#cb3-56"></a>LIRIJP_ExprSet_count <span class="ot">&lt;-</span> <span class="fu">get_ExprSet_LIRIJP</span>(</span>
<span id="cb3-57"><a href="#cb3-57"></a>  <span class="at">profile =</span> LIRIJP_count_data,</span>
<span id="cb3-58"><a href="#cb3-58"></a>  <span class="at">metadata =</span> LIRIJP_phenotype,</span>
<span id="cb3-59"><a href="#cb3-59"></a>  <span class="at">occurrence =</span> <span class="fl">0.2</span>)</span>
<span id="cb3-60"><a href="#cb3-60"></a></span>
<span id="cb3-61"><a href="#cb3-61"></a>LIRIJP_ExprSet_rela <span class="ot">&lt;-</span> <span class="fu">get_ExprSet_LIRIJP</span>(</span>
<span id="cb3-62"><a href="#cb3-62"></a>  <span class="at">profile =</span> LIRIJP_relative_data,</span>
<span id="cb3-63"><a href="#cb3-63"></a>  <span class="at">metadata =</span> LIRIJP_phenotype,</span>
<span id="cb3-64"><a href="#cb3-64"></a>  <span class="at">occurrence =</span> <span class="fl">0.2</span>)</span>
<span id="cb3-65"><a href="#cb3-65"></a>LIRIJP_ExprSet_rela</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="sec-expressionset-LIRIJP-output" class="level3" data-number="8.1.4"><h3 data-number="8.1.4" class="anchored" data-anchor-id="sec-expressionset-LIRIJP-output">
<span class="header-section-number">8.1.4</span> 输出结果</h3>
<p>输出结果: 将上述结果以RDS格式输出到结果目录</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="st">"./data/result/ExpSetObject/"</span>)) {</span>
<span id="cb4-2"><a href="#cb4-2"></a>  <span class="fu">dir.create</span>(<span class="st">"./data/result/ExpSetObject/"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb4-3"><a href="#cb4-3"></a>}</span>
<span id="cb4-4"><a href="#cb4-4"></a></span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="fu">saveRDS</span>(LIRIJP_ExprSet_count, <span class="st">"./data/result/ExpSetObject/LIRI-JP_ExpSet_counts.RDS"</span>, <span class="at">compress =</span> <span class="cn">TRUE</span>)</span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="fu">saveRDS</span>(LIRIJP_ExprSet_rela, <span class="st">"./data/result/ExpSetObject/LIRI-JP_ExpSet_relative.RDS"</span>, <span class="at">compress =</span> <span class="cn">TRUE</span>)</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section></section><section id="sec-expressionset-TCGALIHC" class="level2" data-number="8.2"><h2 data-number="8.2" class="anchored" data-anchor-id="sec-expressionset-TCGALIHC">
<span class="header-section-number">8.2</span> TCGA-LIHC</h2>
<p>要生成<code>TCGA-LIHC</code>的<em>ExpressionSet</em>对象，首先需要加载必要的R包，然后导入数据和元数据，最后将这些数据组合成一个<em>ExpressionSet</em>对象。</p>
<section id="sec-expressionset-TCGALIHC-package" class="level3" data-number="8.2.1"><h3 data-number="8.2.1" class="anchored" data-anchor-id="sec-expressionset-TCGALIHC-package">
<span class="header-section-number">8.2.1</span> 加载R包</h3>
<p>加载R包：使用<code>rm(list = ls())</code>清空环境内变量</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="fu">library</span>(data.table)</span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="fu">library</span>(Biobase)</span>
<span id="cb5-4"><a href="#cb5-4"></a></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="fu">options</span>(<span class="at">stringsAsFactors =</span> F)</span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize =</span> <span class="dv">10000</span> <span class="sc">*</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb5-8"><a href="#cb5-8"></a></span>
<span id="cb5-9"><a href="#cb5-9"></a>grp_names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Early Stage"</span>, <span class="st">"Late Stage"</span>)</span>
<span id="cb5-10"><a href="#cb5-10"></a>grp_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#1F78B4"</span>, <span class="st">"#EE2B2B"</span>)</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="sec-expressionset-TCGALIHC-importdata" class="level3" data-number="8.2.2"><h3 data-number="8.2.2" class="anchored" data-anchor-id="sec-expressionset-TCGALIHC-importdata">
<span class="header-section-number">8.2.2</span> 导入数据</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a></span>
<span id="cb6-2"><a href="#cb6-2"></a>TCGALIHC_count_data <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"./data/result/profile/TCGA-LIHC_gene_counts.tsv"</span>)</span>
<span id="cb6-3"><a href="#cb6-3"></a>TCGALIHC_TrueCount_data <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"./data/result/profile/TCGA-LIHC_gene_Truecounts.tsv"</span>)</span>
<span id="cb6-4"><a href="#cb6-4"></a>TCGALIHC_fpkm_data <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"./data/result/profile/TCGA-LIHC_gene_fpkm.tsv"</span>) </span>
<span id="cb6-5"><a href="#cb6-5"></a>TCGALIHC_phenotype <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"./data/result/metadata/TCGA-LIHC-clinical.csv"</span>)</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="sec-expressionset-TCGALIHC-ExpressionSet" class="level3" data-number="8.2.3"><h3 data-number="8.2.3" class="anchored" data-anchor-id="sec-expressionset-TCGALIHC-ExpressionSet">
<span class="header-section-number">8.2.3</span> 生成ExpressionSet</h3>
<p>生成<em>ExpressionSet-class</em>数据对象：过滤掉低出现率的基因，默认出现率阈值是<strong>20%</strong></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a></span>
<span id="cb7-2"><a href="#cb7-2"></a>get_ExprSet_TCGALIHC <span class="ot">&lt;-</span> <span class="cf">function</span>(</span>
<span id="cb7-3"><a href="#cb7-3"></a>    profile,</span>
<span id="cb7-4"><a href="#cb7-4"></a>    metadata,</span>
<span id="cb7-5"><a href="#cb7-5"></a>    <span class="at">occurrence =</span> <span class="fl">0.2</span>) {</span>
<span id="cb7-6"><a href="#cb7-6"></a>  </span>
<span id="cb7-7"><a href="#cb7-7"></a>  profile_new <span class="ot">&lt;-</span> profile <span class="sc">%&gt;%</span></span>
<span id="cb7-8"><a href="#cb7-8"></a>    tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"V1"</span>)</span>
<span id="cb7-9"><a href="#cb7-9"></a>  metadata_new <span class="ot">&lt;-</span> metadata <span class="sc">%&gt;%</span></span>
<span id="cb7-10"><a href="#cb7-10"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(Tumour_Stage <span class="sc">!=</span> <span class="st">""</span>) <span class="sc">%&gt;%</span></span>
<span id="cb7-11"><a href="#cb7-11"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">Group =</span> <span class="fu">case_when</span>(</span>
<span id="cb7-12"><a href="#cb7-12"></a>      Tumour_Stage <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"T1"</span>, <span class="st">"T2"</span>) <span class="sc">~</span> <span class="st">"Early Stage"</span>, </span>
<span id="cb7-13"><a href="#cb7-13"></a>      Tumour_Stage <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"T3"</span>, <span class="st">"T4"</span>) <span class="sc">~</span> <span class="st">"Late Stage"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb7-14"><a href="#cb7-14"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">Time =</span> <span class="fu">ifelse</span>(<span class="fu">is.na</span>(death_Time), follow_up_Time, death_Time))</span>
<span id="cb7-15"><a href="#cb7-15"></a>  </span>
<span id="cb7-16"><a href="#cb7-16"></a>  <span class="co"># metadata Description</span></span>
<span id="cb7-17"><a href="#cb7-17"></a>  sid <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">intersect</span>(<span class="fu">colnames</span>(profile_new), metadata_new<span class="sc">$</span>PatientID)</span>
<span id="cb7-18"><a href="#cb7-18"></a>  phen <span class="ot">&lt;-</span> metadata_new[<span class="fu">pmatch</span>(sid, metadata_new<span class="sc">$</span>PatientID), , ]</span>
<span id="cb7-19"><a href="#cb7-19"></a>  prof <span class="ot">&lt;-</span> profile_new <span class="sc">%&gt;%</span></span>
<span id="cb7-20"><a href="#cb7-20"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(dplyr<span class="sc">::</span><span class="fu">all_of</span>(phen<span class="sc">$</span>PatientID))</span>
<span id="cb7-21"><a href="#cb7-21"></a>  </span>
<span id="cb7-22"><a href="#cb7-22"></a>  <span class="co"># filter features according gene symbol</span></span>
<span id="cb7-23"><a href="#cb7-23"></a>  prof_cln <span class="ot">&lt;-</span> prof <span class="sc">%&gt;%</span></span>
<span id="cb7-24"><a href="#cb7-24"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"Gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb7-25"><a href="#cb7-25"></a>    <span class="fu">data.frame</span>()</span>
<span id="cb7-26"><a href="#cb7-26"></a>  idx <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"Gene"</span>, <span class="fu">colnames</span>(prof_cln))</span>
<span id="cb7-27"><a href="#cb7-27"></a>  prof_cln<span class="sc">$</span>median <span class="ot">&lt;-</span> <span class="fu">apply</span>(prof_cln[, <span class="sc">-</span>idx], <span class="dv">1</span>, median)</span>
<span id="cb7-28"><a href="#cb7-28"></a>  prof_cln <span class="ot">&lt;-</span> <span class="fu">with</span>(prof_cln, prof_cln[<span class="fu">order</span>(Gene, median, <span class="at">decreasing =</span> T), ])</span>
<span id="cb7-29"><a href="#cb7-29"></a>  prf_deduplicated <span class="ot">&lt;-</span> prof_cln[<span class="sc">!</span><span class="fu">duplicated</span>(prof_cln<span class="sc">$</span>Gene), ] <span class="sc">%&gt;%</span></span>
<span id="cb7-30"><a href="#cb7-30"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>median)</span>
<span id="cb7-31"><a href="#cb7-31"></a>  </span>
<span id="cb7-32"><a href="#cb7-32"></a>  <span class="co"># return result: gene symbol</span></span>
<span id="cb7-33"><a href="#cb7-33"></a>  <span class="fu">rownames</span>(prf_deduplicated) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb7-34"><a href="#cb7-34"></a>  prof_res <span class="ot">&lt;-</span> prf_deduplicated <span class="sc">%&gt;%</span></span>
<span id="cb7-35"><a href="#cb7-35"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(Gene, <span class="fu">everything</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb7-36"><a href="#cb7-36"></a>    tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"Gene"</span>)</span>
<span id="cb7-37"><a href="#cb7-37"></a>  </span>
<span id="cb7-38"><a href="#cb7-38"></a>  <span class="co"># determine the right order between profile and phenotype</span></span>
<span id="cb7-39"><a href="#cb7-39"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(prof_trim)) {</span>
<span id="cb7-40"><a href="#cb7-40"></a>    <span class="cf">if</span> (<span class="sc">!</span>(<span class="fu">colnames</span>(prof_trim)[i] <span class="sc">==</span> <span class="fu">rownames</span>(phen)[i])) {</span>
<span id="cb7-41"><a href="#cb7-41"></a>      <span class="fu">stop</span>(<span class="fu">paste0</span>(i, <span class="st">" Wrong"</span>))</span>
<span id="cb7-42"><a href="#cb7-42"></a>    }</span>
<span id="cb7-43"><a href="#cb7-43"></a>  }</span>
<span id="cb7-44"><a href="#cb7-44"></a></span>
<span id="cb7-45"><a href="#cb7-45"></a>  expressionSet <span class="ot">&lt;-</span> <span class="fu">new</span>(<span class="st">"ExpressionSet"</span>, </span>
<span id="cb7-46"><a href="#cb7-46"></a>                       <span class="at">exprs =</span> exprs,</span>
<span id="cb7-47"><a href="#cb7-47"></a>                       <span class="at">phenoData =</span> adf,</span>
<span id="cb7-48"><a href="#cb7-48"></a>                       <span class="at">experimentData =</span> experimentData)</span>
<span id="cb7-49"><a href="#cb7-49"></a>  </span>
<span id="cb7-50"><a href="#cb7-50"></a>  <span class="fu">return</span>(expressionSet)</span>
<span id="cb7-51"><a href="#cb7-51"></a>}</span>
<span id="cb7-52"><a href="#cb7-52"></a></span>
<span id="cb7-53"><a href="#cb7-53"></a>TCGALIHC_ExprSet_count <span class="ot">&lt;-</span> <span class="fu">get_ExprSet_TCGALIHC</span>(</span>
<span id="cb7-54"><a href="#cb7-54"></a>  <span class="at">profile =</span> TCGALIHC_count_data,</span>
<span id="cb7-55"><a href="#cb7-55"></a>  <span class="at">metadata =</span> TCGALIHC_phenotype,</span>
<span id="cb7-56"><a href="#cb7-56"></a>  <span class="at">occurrence =</span> <span class="fl">0.2</span>)</span>
<span id="cb7-57"><a href="#cb7-57"></a><span class="co"># TCGALIHC_ExprSet_count</span></span>
<span id="cb7-58"><a href="#cb7-58"></a></span>
<span id="cb7-59"><a href="#cb7-59"></a>TCGALIHC_ExprSet_TrueCount <span class="ot">&lt;-</span> <span class="fu">get_ExprSet_TCGALIHC</span>(</span>
<span id="cb7-60"><a href="#cb7-60"></a>  <span class="at">profile =</span> TCGALIHC_TrueCount_data,</span>
<span id="cb7-61"><a href="#cb7-61"></a>  <span class="at">metadata =</span> TCGALIHC_phenotype,</span>
<span id="cb7-62"><a href="#cb7-62"></a>  <span class="at">occurrence =</span> <span class="fl">0.2</span>)</span>
<span id="cb7-63"><a href="#cb7-63"></a><span class="co"># TCGALIHC_ExprSet_TrueCount</span></span>
<span id="cb7-64"><a href="#cb7-64"></a></span>
<span id="cb7-65"><a href="#cb7-65"></a>TCGALIHC_ExprSet_fpkm <span class="ot">&lt;-</span> <span class="fu">get_ExprSet_TCGALIHC</span>(</span>
<span id="cb7-66"><a href="#cb7-66"></a>  <span class="at">profile =</span> TCGALIHC_fpkm_data,</span>
<span id="cb7-67"><a href="#cb7-67"></a>  <span class="at">metadata =</span> TCGALIHC_phenotype,</span>
<span id="cb7-68"><a href="#cb7-68"></a>  <span class="at">occurrence =</span> <span class="fl">0.2</span>)</span>
<span id="cb7-69"><a href="#cb7-69"></a>TCGALIHC_ExprSet_fpkm</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="sec-expressionset-TCGALIHC-output" class="level3" data-number="8.2.4"><h3 data-number="8.2.4" class="anchored" data-anchor-id="sec-expressionset-TCGALIHC-output">
<span class="header-section-number">8.2.4</span> 输出结果</h3>
<p>输出结果: 将上述结果以RDS格式输出到结果目录</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a></span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="st">"./data/result/ExpSetObject/"</span>)) {</span>
<span id="cb8-3"><a href="#cb8-3"></a>  <span class="fu">dir.create</span>(<span class="st">"./data/result/ExpSetObject/"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-4"><a href="#cb8-4"></a>}</span>
<span id="cb8-5"><a href="#cb8-5"></a></span>
<span id="cb8-6"><a href="#cb8-6"></a><span class="fu">saveRDS</span>(TCGALIHC_ExprSet_count, <span class="st">"./data/result/ExpSetObject/TCGA-LIHC_ExpSet_counts.RDS"</span>, <span class="at">compress =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="fu">saveRDS</span>(TCGALIHC_ExprSet_TrueCount, <span class="st">"./data/result/ExpSetObject/TCGA-LIHC_ExpSet_TrueCount.RDS"</span>, <span class="at">compress =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-8"><a href="#cb8-8"></a><span class="fu">saveRDS</span>(TCGALIHC_ExprSet_fpkm, <span class="st">"./data/result/ExpSetObject/TCGA-LIHC_ExpSet_fpkm.RDS"</span>, <span class="at">compress =</span> <span class="cn">TRUE</span>)</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section></section><section id="sec-expressionset-GSE14520" class="level2" data-number="8.3"><h2 data-number="8.3" class="anchored" data-anchor-id="sec-expressionset-GSE14520">
<span class="header-section-number">8.3</span> GSE14520</h2>
<p>要生成<code>GSE14520</code>的<em>ExpressionSet</em>对象，首先需要加载必要的R包，然后导入数据和元数据，最后将这些数据组合成一个<em>ExpressionSet</em>对象。</p>
<section id="sec-expressionset-GSE14520-package" class="level3" data-number="8.3.1"><h3 data-number="8.3.1" class="anchored" data-anchor-id="sec-expressionset-GSE14520-package">
<span class="header-section-number">8.3.1</span> 加载R包</h3>
<p>加载R包：使用<code>rm(list = ls())</code>清空环境内变量</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="fu">library</span>(data.table)</span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="fu">library</span>(Biobase)</span>
<span id="cb9-4"><a href="#cb9-4"></a></span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="fu">options</span>(<span class="at">stringsAsFactors =</span> F)</span>
<span id="cb9-7"><a href="#cb9-7"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize =</span> <span class="dv">10000</span> <span class="sc">*</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb9-8"><a href="#cb9-8"></a></span>
<span id="cb9-9"><a href="#cb9-9"></a>grp_names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Early Stage"</span>, <span class="st">"Late Stage"</span>)</span>
<span id="cb9-10"><a href="#cb9-10"></a>grp_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#1F78B4"</span>, <span class="st">"#EE2B2B"</span>)</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="sec-expressionset-GSE14520-importdata" class="level3" data-number="8.3.2"><h3 data-number="8.3.2" class="anchored" data-anchor-id="sec-expressionset-GSE14520-importdata">
<span class="header-section-number">8.3.2</span> 导入数据</h3>
<p>导入数据</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a></span>
<span id="cb10-2"><a href="#cb10-2"></a>GSE14520_count_data <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"./data/result/profile/GSE14520_gene_counts.tsv"</span>)</span>
<span id="cb10-3"><a href="#cb10-3"></a>GSE14520_phenotype <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"./data/result/metadata/GSE14520-clinical.csv"</span>)</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="sec-expressionset-GSE14520-ExpressionSet" class="level3" data-number="8.3.3"><h3 data-number="8.3.3" class="anchored" data-anchor-id="sec-expressionset-GSE14520-ExpressionSet">
<span class="header-section-number">8.3.3</span> 生成ExpressionSet</h3>
<p>生成<em>ExpressionSet-class</em>数据对象：过滤掉低出现率的基因，默认出现率阈值是<strong>20%</strong></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a></span>
<span id="cb11-2"><a href="#cb11-2"></a>get_ExprSet_GSE14520 <span class="ot">&lt;-</span> <span class="cf">function</span>(</span>
<span id="cb11-3"><a href="#cb11-3"></a>    profile,</span>
<span id="cb11-4"><a href="#cb11-4"></a>    metadata,</span>
<span id="cb11-5"><a href="#cb11-5"></a>    <span class="at">occurrence =</span> <span class="fl">0.2</span>) {</span>
<span id="cb11-6"><a href="#cb11-6"></a>  </span>
<span id="cb11-7"><a href="#cb11-7"></a>  profile_new <span class="ot">&lt;-</span> profile <span class="sc">%&gt;%</span></span>
<span id="cb11-8"><a href="#cb11-8"></a>    tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"V1"</span>)</span>
<span id="cb11-9"><a href="#cb11-9"></a>  metadata_new <span class="ot">&lt;-</span> metadata <span class="sc">%&gt;%</span></span>
<span id="cb11-10"><a href="#cb11-10"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(Tumour_Stage <span class="sc">!=</span> <span class="st">""</span>) <span class="sc">%&gt;%</span></span>
<span id="cb11-11"><a href="#cb11-11"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">Group =</span> <span class="fu">case_when</span>(</span>
<span id="cb11-12"><a href="#cb11-12"></a>      Tumour_Stage <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"T1"</span>, <span class="st">"T2"</span>) <span class="sc">~</span> <span class="st">"Early Stage"</span>, </span>
<span id="cb11-13"><a href="#cb11-13"></a>      Tumour_Stage <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"T3"</span>, <span class="st">"T4"</span>) <span class="sc">~</span> <span class="st">"Late Stage"</span>)) </span>
<span id="cb11-14"><a href="#cb11-14"></a>  </span>
<span id="cb11-15"><a href="#cb11-15"></a>  <span class="co"># metadata Description</span></span>
<span id="cb11-16"><a href="#cb11-16"></a>  sid <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">intersect</span>(<span class="fu">colnames</span>(profile_new), metadata_new<span class="sc">$</span>SampleID)</span>
<span id="cb11-17"><a href="#cb11-17"></a>  phen <span class="ot">&lt;-</span> metadata_new[<span class="fu">pmatch</span>(sid, metadata_new<span class="sc">$</span>SampleID), , ]</span>
<span id="cb11-18"><a href="#cb11-18"></a>  prof <span class="ot">&lt;-</span> profile_new <span class="sc">%&gt;%</span></span>
<span id="cb11-19"><a href="#cb11-19"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(dplyr<span class="sc">::</span><span class="fu">all_of</span>(phen<span class="sc">$</span>SampleID))</span>
<span id="cb11-20"><a href="#cb11-20"></a>  </span>
<span id="cb11-21"><a href="#cb11-21"></a>  </span>
<span id="cb11-22"><a href="#cb11-22"></a>  <span class="co"># return result: gene symbol</span></span>
<span id="cb11-23"><a href="#cb11-23"></a>  <span class="fu">rownames</span>(prf_deduplicated) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb11-24"><a href="#cb11-24"></a>  prof_res <span class="ot">&lt;-</span> prf_deduplicated <span class="sc">%&gt;%</span></span>
<span id="cb11-25"><a href="#cb11-25"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(Gene, <span class="fu">everything</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb11-26"><a href="#cb11-26"></a>    tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"Gene"</span>)</span>
<span id="cb11-27"><a href="#cb11-27"></a>  </span>
<span id="cb11-28"><a href="#cb11-28"></a>  <span class="co"># determine the right order between profile and phenotype</span></span>
<span id="cb11-29"><a href="#cb11-29"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(prof_trim)) {</span>
<span id="cb11-30"><a href="#cb11-30"></a>    <span class="cf">if</span> (<span class="sc">!</span>(<span class="fu">colnames</span>(prof_trim)[i] <span class="sc">==</span> <span class="fu">rownames</span>(phen)[i])) {</span>
<span id="cb11-31"><a href="#cb11-31"></a>      <span class="fu">stop</span>(<span class="fu">paste0</span>(i, <span class="st">" Wrong"</span>))</span>
<span id="cb11-32"><a href="#cb11-32"></a>    }</span>
<span id="cb11-33"><a href="#cb11-33"></a>  }</span>
<span id="cb11-34"><a href="#cb11-34"></a>  </span>
<span id="cb11-35"><a href="#cb11-35"></a>  expressionSet <span class="ot">&lt;-</span> <span class="fu">new</span>(<span class="st">"ExpressionSet"</span>, </span>
<span id="cb11-36"><a href="#cb11-36"></a>                       <span class="at">exprs =</span> exprs,</span>
<span id="cb11-37"><a href="#cb11-37"></a>                       <span class="at">phenoData =</span> adf,</span>
<span id="cb11-38"><a href="#cb11-38"></a>                       <span class="at">experimentData =</span> experimentData)</span>
<span id="cb11-39"><a href="#cb11-39"></a>  </span>
<span id="cb11-40"><a href="#cb11-40"></a>  <span class="fu">return</span>(expressionSet)</span>
<span id="cb11-41"><a href="#cb11-41"></a>}</span>
<span id="cb11-42"><a href="#cb11-42"></a></span>
<span id="cb11-43"><a href="#cb11-43"></a>GSE14520_ExprSet_count <span class="ot">&lt;-</span> <span class="fu">get_ExprSet_GSE14520</span>(</span>
<span id="cb11-44"><a href="#cb11-44"></a>  <span class="at">profile =</span> GSE14520_count_data,</span>
<span id="cb11-45"><a href="#cb11-45"></a>  <span class="at">metadata =</span> GSE14520_phenotype,</span>
<span id="cb11-46"><a href="#cb11-46"></a>  <span class="at">occurrence =</span> <span class="fl">0.2</span>)</span>
<span id="cb11-47"><a href="#cb11-47"></a>GSE14520_ExprSet_count</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="sec-expressionset-GSE14520-output" class="level3" data-number="8.3.4"><h3 data-number="8.3.4" class="anchored" data-anchor-id="sec-expressionset-GSE14520-output">
<span class="header-section-number">8.3.4</span> 输出结果</h3>
<p>输出结果: 将上述结果以RDS格式输出到结果目录</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a></span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="st">"./data/result/ExpSetObject/"</span>)) {</span>
<span id="cb12-3"><a href="#cb12-3"></a>  <span class="fu">dir.create</span>(<span class="st">"./data/result/ExpSetObject/"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-4"><a href="#cb12-4"></a>}</span>
<span id="cb12-5"><a href="#cb12-5"></a></span>
<span id="cb12-6"><a href="#cb12-6"></a><span class="fu">saveRDS</span>(GSE14520_ExprSet_count, <span class="st">"./data/result/ExpSetObject/GSE14520_ExpSet_counts.RDS"</span>, <span class="at">compress =</span> <span class="cn">TRUE</span>)</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section></section><section id="sec-expressionset-overlapgene" class="level2" data-number="8.4"><h2 data-number="8.4" class="anchored" data-anchor-id="sec-expressionset-overlapgene">
<span class="header-section-number">8.4</span> 重叠基因</h2>
<p>为了确保在发现数据集中识别的标记基因能够在验证数据集中进行准确的验证，首先加载了基于<em>count abundance</em>的RDS，这些RDS分别包含<code>LIRI-JP</code>，<code>LIHC-US/TCGA-LIHC</code>和<code>GSE14520</code>数据集的基因表达信息。</p>
<p><strong>步骤</strong>：</p>
<ol type="1">
<li><p>加载RDS文件：首先，将分别加载<code>LIRI-JP</code>，<code>LIHC-US/TCGA-LIHC</code>和<code>GSE14520</code>的<em>count abundance</em>的RDS文件。</p></li>
<li><p>提取基因标识符：从每个数据集中提取基因标识符列表。</p></li>
<li><p>获取基因交集：比较这三个基因标识符列表，找出它们之间的交集。这个交集将包含所有在三个数据集中都存在的基因。</p></li>
<li><p>更新数据集：使用交集中的基因标识符来过滤和更新每个数据集，以确保后续分析中的基因一致性。</p></li>
</ol>
<section id="sec-expressionset-overlapgene-package" class="level3" data-number="8.4.1"><h3 data-number="8.4.1" class="anchored" data-anchor-id="sec-expressionset-overlapgene-package">
<span class="header-section-number">8.4.1</span> 加载R包</h3>
<p>加载R包</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="fu">library</span>(data.table)</span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="fu">library</span>(Biobase)</span>
<span id="cb13-4"><a href="#cb13-4"></a></span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb13-6"><a href="#cb13-6"></a><span class="fu">options</span>(<span class="at">stringsAsFactors =</span> F)</span>
<span id="cb13-7"><a href="#cb13-7"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize =</span> <span class="dv">10000</span> <span class="sc">*</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb13-8"><a href="#cb13-8"></a></span>
<span id="cb13-9"><a href="#cb13-9"></a>grp_names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Early Stage"</span>, <span class="st">"Late Stage"</span>)</span>
<span id="cb13-10"><a href="#cb13-10"></a>grp_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#1F78B4"</span>, <span class="st">"#EE2B2B"</span>)</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="sec-expressionset-overlapgene-importdata" class="level3" data-number="8.4.2"><h3 data-number="8.4.2" class="anchored" data-anchor-id="sec-expressionset-overlapgene-importdata">
<span class="header-section-number">8.4.2</span> 导入数据</h3>
<p>导入数据</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>ExprSet_LIRI_JP <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./data/result/ExpSetObject/LIRI-JP_ExpSet_counts.RDS"</span>)</span>
<span id="cb14-2"><a href="#cb14-2"></a>ExprSet_TCGA_LIHC <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./data/result/ExpSetObject/TCGA-LIHC_ExpSet_TrueCount.RDS"</span>)</span>
<span id="cb14-3"><a href="#cb14-3"></a>ExprSet_GSE14520 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./data/result/ExpSetObject/GSE14520_ExpSet_counts.RDS"</span>)</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="sec-expressionset-overlapgene-common" class="level3" data-number="8.4.3"><h3 data-number="8.4.3" class="anchored" data-anchor-id="sec-expressionset-overlapgene-common">
<span class="header-section-number">8.4.3</span> 共有基因</h3>
<p>共有基因</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a></span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="fu">length</span>(overlap_genes)</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="sec-expressionset-overlapgene-object" class="level3" data-number="8.4.4"><h3 data-number="8.4.4" class="anchored" data-anchor-id="sec-expressionset-overlapgene-object">
<span class="header-section-number">8.4.4</span> 生成数据对象</h3>
<p>生成数据对象</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a></span>
<span id="cb16-2"><a href="#cb16-2"></a>get_ExprSet <span class="ot">&lt;-</span> <span class="cf">function</span>(</span>
<span id="cb16-3"><a href="#cb16-3"></a>    dat,</span>
<span id="cb16-4"><a href="#cb16-4"></a>    <span class="at">gene_list =</span> overlap_genes,</span>
<span id="cb16-5"><a href="#cb16-5"></a>    <span class="at">occurrence =</span> <span class="fl">0.2</span>) {</span>
<span id="cb16-6"><a href="#cb16-6"></a>  </span>
<span id="cb16-7"><a href="#cb16-7"></a>  profile <span class="ot">&lt;-</span> <span class="fu">exprs</span>(dat) <span class="sc">%&gt;%</span></span>
<span id="cb16-8"><a href="#cb16-8"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb16-9"><a href="#cb16-9"></a>    <span class="fu">t</span>() <span class="sc">%&gt;%</span></span>
<span id="cb16-10"><a href="#cb16-10"></a>    <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb16-11"><a href="#cb16-11"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">all_of</span>(overlap_genes)) <span class="sc">%&gt;%</span></span>
<span id="cb16-12"><a href="#cb16-12"></a>    <span class="fu">t</span>() <span class="sc">%&gt;%</span></span>
<span id="cb16-13"><a href="#cb16-13"></a>    <span class="fu">as.data.frame</span>()</span>
<span id="cb16-14"><a href="#cb16-14"></a>  metadata <span class="ot">&lt;-</span> <span class="fu">pData</span>(dat) <span class="sc">%&gt;%</span></span>
<span id="cb16-15"><a href="#cb16-15"></a>    <span class="fu">as.data.frame</span>()</span>
<span id="cb16-16"><a href="#cb16-16"></a>  </span>
<span id="cb16-17"><a href="#cb16-17"></a>  <span class="co"># metadata Description</span></span>
<span id="cb16-18"><a href="#cb16-18"></a>  sid <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">intersect</span>(<span class="fu">colnames</span>(profile), <span class="fu">rownames</span>(metadata))</span>
<span id="cb16-19"><a href="#cb16-19"></a>  phen <span class="ot">&lt;-</span> metadata[<span class="fu">pmatch</span>(sid, <span class="fu">rownames</span>(metadata)), , ]</span>
<span id="cb16-20"><a href="#cb16-20"></a>  prof <span class="ot">&lt;-</span> profile <span class="sc">%&gt;%</span></span>
<span id="cb16-21"><a href="#cb16-21"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(dplyr<span class="sc">::</span><span class="fu">all_of</span>(<span class="fu">rownames</span>(phen)))</span>
<span id="cb16-22"><a href="#cb16-22"></a>  </span>
<span id="cb16-23"><a href="#cb16-23"></a>  <span class="co"># filter features according gene symbol</span></span>
<span id="cb16-24"><a href="#cb16-24"></a>  prof_cln <span class="ot">&lt;-</span> prof <span class="sc">%&gt;%</span></span>
<span id="cb16-25"><a href="#cb16-25"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"Gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb16-26"><a href="#cb16-26"></a>    <span class="fu">data.frame</span>()</span>
<span id="cb16-27"><a href="#cb16-27"></a>  idx <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"Gene"</span>, <span class="fu">colnames</span>(prof_cln))</span>
<span id="cb16-28"><a href="#cb16-28"></a>  prof_cln<span class="sc">$</span>median <span class="ot">&lt;-</span> <span class="fu">apply</span>(prof_cln[, <span class="sc">-</span>idx], <span class="dv">1</span>, median)</span>
<span id="cb16-29"><a href="#cb16-29"></a>  prof_cln <span class="ot">&lt;-</span> <span class="fu">with</span>(prof_cln, prof_cln[<span class="fu">order</span>(Gene, median, <span class="at">decreasing =</span> T), ])</span>
<span id="cb16-30"><a href="#cb16-30"></a>  prf_deduplicated <span class="ot">&lt;-</span> prof_cln[<span class="sc">!</span><span class="fu">duplicated</span>(prof_cln<span class="sc">$</span>Gene), ] <span class="sc">%&gt;%</span></span>
<span id="cb16-31"><a href="#cb16-31"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>median)</span>
<span id="cb16-32"><a href="#cb16-32"></a></span>
<span id="cb16-33"><a href="#cb16-33"></a>  </span>
<span id="cb16-34"><a href="#cb16-34"></a>  <span class="co"># determine the right order between profile and phenotype</span></span>
<span id="cb16-35"><a href="#cb16-35"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(prof_trim)) {</span>
<span id="cb16-36"><a href="#cb16-36"></a>    <span class="cf">if</span> (<span class="sc">!</span>(<span class="fu">colnames</span>(prof_trim)[i] <span class="sc">==</span> <span class="fu">rownames</span>(phen)[i])) {</span>
<span id="cb16-37"><a href="#cb16-37"></a>      <span class="fu">stop</span>(<span class="fu">paste0</span>(i, <span class="st">" Wrong"</span>))</span>
<span id="cb16-38"><a href="#cb16-38"></a>    }</span>
<span id="cb16-39"><a href="#cb16-39"></a>  }</span>
<span id="cb16-40"><a href="#cb16-40"></a>  expressionSet <span class="ot">&lt;-</span> <span class="fu">new</span>(<span class="st">"ExpressionSet"</span>, </span>
<span id="cb16-41"><a href="#cb16-41"></a>                       <span class="at">exprs =</span> exprs,</span>
<span id="cb16-42"><a href="#cb16-42"></a>                       <span class="at">phenoData =</span> adf,</span>
<span id="cb16-43"><a href="#cb16-43"></a>                       <span class="at">experimentData =</span> experimentData)</span>
<span id="cb16-44"><a href="#cb16-44"></a>  </span>
<span id="cb16-45"><a href="#cb16-45"></a>  <span class="fu">return</span>(expressionSet)</span>
<span id="cb16-46"><a href="#cb16-46"></a>}</span>
<span id="cb16-47"><a href="#cb16-47"></a></span>
<span id="cb16-48"><a href="#cb16-48"></a></span>
<span id="cb16-49"><a href="#cb16-49"></a>ExprSet_LIRI_JP_new <span class="ot">&lt;-</span> <span class="fu">get_ExprSet</span>(</span>
<span id="cb16-50"><a href="#cb16-50"></a>  <span class="at">dat =</span> ExprSet_LIRI_JP,</span>
<span id="cb16-51"><a href="#cb16-51"></a>  <span class="at">gene_list =</span> overlap_genes,</span>
<span id="cb16-52"><a href="#cb16-52"></a>  <span class="at">occurrence =</span> <span class="fl">0.2</span>)</span>
<span id="cb16-53"><a href="#cb16-53"></a>ExprSet_LIRI_JP_new</span>
<span id="cb16-54"><a href="#cb16-54"></a></span>
<span id="cb16-55"><a href="#cb16-55"></a>ExprSet_TCGA_LIHC_new <span class="ot">&lt;-</span> <span class="fu">get_ExprSet</span>(</span>
<span id="cb16-56"><a href="#cb16-56"></a>  <span class="at">dat =</span> ExprSet_TCGA_LIHC,</span>
<span id="cb16-57"><a href="#cb16-57"></a>  <span class="at">gene_list =</span> overlap_genes,</span>
<span id="cb16-58"><a href="#cb16-58"></a>  <span class="at">occurrence =</span> <span class="fl">0.2</span>)</span>
<span id="cb16-59"><a href="#cb16-59"></a>ExprSet_TCGA_LIHC_new</span>
<span id="cb16-60"><a href="#cb16-60"></a></span>
<span id="cb16-61"><a href="#cb16-61"></a>ExprSet_GSE14520_new <span class="ot">&lt;-</span> <span class="fu">get_ExprSet</span>(</span>
<span id="cb16-62"><a href="#cb16-62"></a>  <span class="at">dat =</span> ExprSet_GSE14520,</span>
<span id="cb16-63"><a href="#cb16-63"></a>  <span class="at">gene_list =</span> overlap_genes,</span>
<span id="cb16-64"><a href="#cb16-64"></a>  <span class="at">occurrence =</span> <span class="fl">0.2</span>)</span>
<span id="cb16-65"><a href="#cb16-65"></a>ExprSet_GSE14520_new</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section><section id="sec-expressionset-overlapgene-output" class="level3" data-number="8.4.5"><h3 data-number="8.4.5" class="anchored" data-anchor-id="sec-expressionset-overlapgene-output">
<span class="header-section-number">8.4.5</span> 输出结果</h3>
<p>输出结果: 将上述结果以RDS格式输出到结果目录</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a></span>
<span id="cb17-2"><a href="#cb17-2"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="st">"./data/result/ExpSetObject/"</span>)) {</span>
<span id="cb17-3"><a href="#cb17-3"></a>  <span class="fu">dir.create</span>(<span class="st">"./data/result/ExpSetObject/"</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-4"><a href="#cb17-4"></a>}</span>
<span id="cb17-5"><a href="#cb17-5"></a></span>
<span id="cb17-6"><a href="#cb17-6"></a><span class="fu">saveRDS</span>(ExprSet_LIRI_JP_new, <span class="st">"./data/result/ExpSetObject/Final_ExpSet_LIRI_JP_TrueCounts.RDS"</span>, <span class="at">compress =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-7"><a href="#cb17-7"></a><span class="fu">saveRDS</span>(ExprSet_TCGA_LIHC_new, <span class="st">"./data/result/ExpSetObject/Final_ExpSet_TCGA_LIHC_TrueCounts.RDS"</span>, <span class="at">compress =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-8"><a href="#cb17-8"></a><span class="fu">saveRDS</span>(ExprSet_GSE14520_new, <span class="st">"./data/result/ExpSetObject/Final_ExpSet_GSE14520_TrueCounts.RDS"</span>, <span class="at">compress =</span> <span class="cn">TRUE</span>)</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section></section><section id="sec-expressionset-summary" class="level2" data-number="8.5"><h2 data-number="8.5" class="anchored" data-anchor-id="sec-expressionset-summary">
<span class="header-section-number">8.5</span> 总结</h2>
<p>在完成了对三个肝细胞癌（HCC）群体的转录组和临床数据的预处理后，分别将这些数据转换为了<em>ExpressionSet</em>数据对象，并保存为<code>RDS</code>文件格式，以便于后续的数据管理和分析。为了确保后续分析的一致性和可靠性，进一步从这些RDS文件中提取了三个数据集中共有的基因。这些共有的基因是后续分析的基础，因为它们代表了三个群体中均存在并可能参与HCC发展的生物学过程。</p>
<details><summary><strong>系统信息</strong>
</summary><div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="复制到剪贴板" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.3 (2024-02-29)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.2

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Asia/Shanghai
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] Biobase_2.62.0      BiocGenerics_0.48.1 data.table_1.15.4  
 [4] lubridate_1.9.3     forcats_1.0.0       stringr_1.5.1      
 [7] dplyr_1.1.4         purrr_1.0.2         readr_2.1.5        
[10] tidyr_1.3.1         tibble_3.2.1        ggplot2_3.5.1      
[13] tidyverse_2.0.0    

loaded via a namespace (and not attached):
 [1] gtable_0.3.5        jsonlite_1.8.8      compiler_4.3.3     
 [4] BiocManager_1.30.23 renv_1.0.0          tidyselect_1.2.1   
 [7] scales_1.3.0        yaml_2.3.8          fastmap_1.1.1      
[10] R6_2.5.1            generics_0.1.3      knitr_1.46         
[13] htmlwidgets_1.6.4   munsell_0.5.1       tzdb_0.4.0         
[16] pillar_1.9.0        rlang_1.1.3         utf8_1.2.4         
[19] stringi_1.8.4       xfun_0.43           timechange_0.3.0   
[22] cli_3.6.2           withr_3.0.0         magrittr_2.0.3     
[25] digest_0.6.35       grid_4.3.3          rstudioapi_0.16.0  
[28] hms_1.1.3           lifecycle_1.0.4     vctrs_0.6.5        
[31] evaluate_0.23       glue_1.7.0          fansi_1.0.6        
[34] colorspace_2.1-0    rmarkdown_2.26      tools_4.3.3        
[37] pkgconfig_2.0.3     htmltools_0.5.8.1  </code></pre>
</div>
</div>
</details>

<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-huber2015orchestrating" class="csl-entry" role="listitem">
Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, 等. 2015. <span>《Orchestrating high-throughput genomic analysis with Bioconductor》</span>. <em>Nature methods</em> 12 (2): 115–21.
</div>
</div>
</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "已复制");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "已复制");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./wrangling-processing.html" class="pagination-link" aria-label="数据预处理">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">数据预处理</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./wrangling-batchEffect.html" class="pagination-link" aria-label="数据校正">
        <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">数据校正</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/blob/main/wrangling-objects.qmd" class="toc-action"><i class="bi bi-github"></i>查看代码</a></li></ul></div></div></div></footer></body></html>